CustomCraftSML/Serialization/Entries/CustomFood.cs

```diff
@@ -0,0 +1,414 @@
+        };
+
+
+        internal CustomFood(TechType origTechType) : this()
+        {
+            ITechData origRecipe = CraftData.Get(origTechType);
+            this.ItemID = origTechType.ToString();
+            this.AmountCrafted = (short)origRecipe.craftAmount;
+
+            for (int i = 0; i < origRecipe.ingredientCount; i++)
+            {
+                IIngredient origIngredient = origRecipe.GetIngredient(i);
+                this.Ingredients.Add(new EmIngredient(origIngredient.techType, (short)origIngredient.amount));
+            }
+
+            for (int i = 0; i < origRecipe.linkedItemCount; i++)
+                emLinked.Add(origRecipe.GetLinkedItem(i).ToString());
+        }
+
+        public CustomFood() : this(TypeName, CustomFoodProperties)
+        {
+        }
+
+        protected CustomFood(string key) : this(key, CustomFoodProperties)
+        {
+        }
+
+        protected CustomFood(string key, ICollection<EmProperty> definitions) : base(key, definitions)
+        {
+            emAmount = (EmProperty<short>) Properties[AmountCraftedKey];
+            emForceAtStart = (EmYesNo) Properties[ForceUnlockKey];
+            emIngredient = (EmPropertyCollectionList<EmIngredient>) Properties[IngredientsKey];
+            emLinked = (EmPropertyList<string>) Properties[LinkedKey];
+            emUnlocks = (EmPropertyList<string>) Properties[UnlocksKey];
+            emUnlockedby = (EmPropertyList<string>) Properties[UnlockedbyKey];
+            emDisplayname = (EmProperty<string>) Properties[DisplayNameKey];
+            emTooltip = (EmProperty<string>) Properties[TooltipKey];
+            emSpriteID = (EmProperty<TechType>) Properties[SpriteKey];
+            emPath = (EmProperty<string>) Properties[PathKey];
+            emPdagroup = (EmProperty<TechGroup>) Properties[PdagroupKey];
+            emPdacategory = (EmProperty<TechCategory>) Properties[PdacategoryKey];
+            emFood = (EmProperty<short>) Properties[FoodKey];
+            emWater = (EmProperty<short>) Properties[WaterKey];
+            //emDecomposes = (EmYesNo) Properties[DecomposesKey];
+            emDecayrate = (EmProperty<short>) Properties[DecayrateKey];
+            emOverfill = (EmYesNo) Properties[OverfillKey];
+
+            OnValueExtractedEvent += ValueExtracted;
+        }
+
+        private void ValueExtracted()
+        {
+            foreach (EmIngredient ingredient in emIngredient)
+            {
+                string itemID = (ingredient["ItemID"] as EmProperty<string>).Value;
+                short required = (ingredient["Required"] as EmProperty<short>).Value;
+            }
+        }
+
+        internal override EmProperty Copy() => new CustomFood(this.Key, this.CopyDefinitions);
+
+
+
+        public bool SendToSMLHelper()
+        {
+            try
+            {
+                ValidateIngredients();
+                ValidateLinkedItems();
+                ValidateUnlockedBy();
+                ValidateUnlocks();
```

Comment:
All these **Validate** methods should be called from a `PassesPreValidation()` method.  
When returning their value, use `&` instead of `&&` so that all of the methods validate and log any errors found.